### PR TITLE
txn: make kv pairs converted from locks invisible

### PR DESCRIPTION
**Pull request by zyguan**
### What problem does this PR solve?

Issue Number: ref #28011 

Problem Summary: as described in #28011 , kv pairs converted from locks is visible to union-scan, which leads to duplicated rows (one from snapshot and the other from mem-index-reader).

### What is changed and how it works?

Mark keys converted from lock as invisible and bypass those keys on read. Since this PR is only a hotfix for #28011 , it's better not to introduce extra complexity. Thus, instead of adding a new flag to client-go, an invisible key set is added to `tikvTxn` for tracking key visibility.

For v6.5.0 and above, https://github.com/tikv/tikv/issues/13694 is a better solution.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by cfzjywxk — APPROVED** [review]

File: store/driver/txn/unionstore_driver.go [hunk 1]
```diff
@@ -19,41 +19,73 @@ import (
...
 	"github.com/pingcap/tidb/kv"
 	derr "github.com/pingcap/tidb/store/driver/error"
+	tikverr "github.com/tikv/client-go/v2/error"
 	tikvstore "github.com/tikv/client-go/v2/kv"
 	"github.com/tikv/client-go/v2/tikv"
+	"github.com/tikv/client-go/v2/util"
 )
 
+type visibilityChecker interface {
+	isVisible(kv.Key) bool
```
Comment (ekexium) [inline store/driver/txn/unionstore_driver.go:29]: The name is a bit vague. The status is 3-valued: (1)exist + visible, (2)exist + invisible, (3)absent(unknown). 
It returns true iff (1) || (3).

**Review by ekexium — APPROVED** [review]
The rest LGTM.

**Review by you06 — APPROVED** [review]

File: store/driver/txn/unionstore_driver.go (hunk 1 above)
Comment (zyguan) [reply store/driver/txn/unionstore_driver.go:29]: How about to rename it to `invisible`?

Comment (ekexium) [reply store/driver/txn/unionstore_driver.go:29]: Sounds good. Or something like `isVisibleOrUnknown`(?)

**Review by cfzjywxk — DISMISSED** [review]

**Review by MyonKeminta — APPROVED** [review]

**Pull request merged**